### PR TITLE
Validation check for is_callable

### PR DESCRIPTION
**Pull request by imshashank**
Fix for Issue : https://github.com/aws/aws-sdk-php/issues/1257

### CODE REVIEW

**Comment by kstich** [comment]
Strings can be valid callables, we should see if we can perform a better check for this.

File: src/ClientResolver.php [hunk 1]
```diff
@@ -252,6 +252,7 @@ public function resolve(array $args, HandlerList $list)
                 if (isset($a['default'])) {
                     // Merge defaults in when not present.
                     $args[$key] = is_callable($a['default'])
+                    && !is_string($a['default'])
```
Comment (kstich) [inline src/ClientResolver.php]: What about `in_array('callable', $a['valid'])` instead for this check?

Comment (imshashank) [inline src/ClientResolver.php]: That seems like a more appropriate fix.

File: Makefile [hunk 2]
```diff
@@ -1,165 +0,0 @@
-help:
```
Comment (mtdowling) [inline Makefile]: Why is this being removed?

File: src/ClientResolver.php [hunk 3]
```diff
@@ -252,7 +252,9 @@ public function resolve(array $args, HandlerList $list)
                 if (isset($a['default'])) {
                     // Merge defaults in when not present.
                     $args[$key] = is_callable($a['default'])
-                    && !is_string($a['default'])
+                    && (in_array('callable', $a['valid'])
```
Comment (mtdowling) [inline src/ClientResolver.php]: Let's break this out from a ternary expression into some more readable like if statements.

**Comment by imshashank** [comment]
@mtdowling @kstich Please check this and let me know if this looks good. Thanks

File: tests/ClientResolverTest.php [hunk 4]
```diff
@@ -167,6 +167,19 @@ public function testValidatesCallables()
...
+            'foo' => [
+                'type'   => 'value',
+                'valid'  => ['string'],
+                'default' => function (){return 'callable_test';}
+            ]
+        ]);
+        $res = $r->resolve([], new HandlerList());
+        $this->assertEquals('callable_test',$res['foo']);
+    }
+
```
Comment (kstich) [inline tests/ClientResolverTest.php]: Should also add a test that a string default that is a parameter-less function (say, `gc_collect_cycles`) is not invoked.

Comment (imshashank) [inline tests/ClientResolverTest.php]: Added more test

**Comment by mtdowling** [comment]
:ship:

**Pull request merged**